extensions/mailer/runtime/src/main/java/io/quarkus/mailer/runtime/MailClientProducer.java

```diff
@@ -16,52 +20,74 @@
+        this.client = mailClient(vertx, config);
+        this.mutinyClient = io.vertx.mutiny.ext.mail.MailClient.newInstance(this.client);
     }
 
     @Singleton
     @Produces
-    public synchronized MailClient mailClient() {
+    public MailClient mailClient() {
         return client;
     }
 
     @Singleton
     @Produces
-    public synchronized io.vertx.mutiny.ext.mail.MailClient mutinyClient() {
+    public io.vertx.mutiny.ext.mail.MailClient mutinyClient() {
         return mutinyClient;
     }
 
     @Singleton
     @Produces
     @Deprecated
-    public synchronized io.vertx.axle.ext.mail.MailClient axleMailClient() {
-        if (axleMailClient == null) {
-            LOGGER.warn(
-                    "`io.vertx.axle.ext.mail.MailClient` is deprecated and will be removed in a future version - it is "
-                            + "recommended to switch to `io.vertx.mutiny.ext.mail.MailClient`");
-            axleMailClient = io.vertx.axle.ext.mail.MailClient.newInstance(client);
-        }
-        return axleMailClient;
+    public io.vertx.axle.ext.mail.MailClient axleMailClient() {
+        LOGGER.warn(
+                "`io.vertx.axle.ext.mail.MailClient` is deprecated and will be removed in a future version - it is "
+                        + "recommended to switch to `io.vertx.mutiny.ext.mail.MailClient`");
+        return io.vertx.axle.ext.mail.MailClient.newInstance(client);
     }
 
     @Singleton
     @Produces
     @Deprecated
-    public synchronized io.vertx.reactivex.ext.mail.MailClient rxMailClient() {
-        if (rxMailClient == null) {
-            LOGGER.warn(
-                    "`io.vertx.reactivex.ext.mail.MailClient` is deprecated and will be removed in a future version - it is "
-                            + "recommended to switch to `io.vertx.mutiny.ext.mail.MailClient`");
-            rxMailClient = io.vertx.reactivex.ext.mail.MailClient.newInstance(client);
-        }
-        return rxMailClient;
+    public io.vertx.reactivex.ext.mail.MailClient rxMailClient() {
+        LOGGER.warn(
+                "`io.vertx.reactivex.ext.mail.MailClient` is deprecated and will be removed in a future version - it is "
+                        + "recommended to switch to `io.vertx.mutiny.ext.mail.MailClient`");
+        return io.vertx.reactivex.ext.mail.MailClient.newInstance(client);
+    }
+
+    @PreDestroy
+    public void stop() {
+        client.close();
+    }
+
+    private MailClient mailClient(Vertx vertx, MailConfig config) {
+        io.vertx.ext.mail.MailConfig cfg = toVertxMailConfig(config);
+        return MailClient.createShared(vertx, cfg);
+    }
+
+    private io.vertx.ext.mail.MailConfig toVertxMailConfig(MailConfig config) {
+        io.vertx.ext.mail.MailConfig cfg = new io.vertx.ext.mail.MailConfig();
+        config.authMethods.ifPresent(cfg::setAuthMethods);
+        cfg.setDisableEsmtp(config.disableEsmtp);
+        cfg.setHostname(config.host);
```

Comment:
So when you add the mailer extension, this is mandatory and things will fail if undefined? It's a bit different than for the other extensions.

Not related to this patch but I wonder if we should revisit this at some point.

Reply:
I would prefer not touch that now TBH

Reply:
Yeah agreed, thus the "at some point". We probably need to discuss that before changing anything.

Reply:
:+1: